internal/controller/controllers/agentserviceconfig_controller.go

```diff
@@ -0,0 +1,671 @@
+			// Return and don't requeue
+			return reconcile.Result{}, nil
+		}
+		r.Log.Error(err, "Failed to get resource", req.NamespacedName)
+		return ctrl.Result{}, err
+	}
+
+	for _, f := range []func(context.Context, *adiiov1alpha1.AgentServiceConfig) error{
+		r.ensureFilesystemStorage,
+		r.ensureDatabaseStorage,
+		r.ensureAgentService,
+		r.ensureAgentRoute,
+		r.ensurePostgresSecret,
+		r.ensureAssistedServiceDeployment,
+	} {
+		err := f(ctx, instance)
+		if err != nil {
+			r.Log.Error(err, "Failed reconcile")
+			if statusErr := r.Status().Update(ctx, instance); statusErr != nil {
+				r.Log.Error(err, "Failed to update status")
+				return ctrl.Result{Requeue: true}, statusErr
+			}
+			return ctrl.Result{Requeue: true}, err
+		}
+	}
+
+	conditionsv1.SetStatusCondition(&instance.Status.Conditions, conditionsv1.Condition{
+		Type:    adiiov1alpha1.ConditionReconcileCompleted,
+		Status:  corev1.ConditionTrue,
+		Reason:  "ReconcileCompleted",
+		Message: "AgentServiceConfig reconcile completed without error.",
+	})
+	return ctrl.Result{}, r.Status().Update(ctx, instance)
+}
+
+func (r *AgentServiceConfigReconciler) ensureFilesystemStorage(ctx context.Context, instance *adiiov1alpha1.AgentServiceConfig) error {
+	pvc, mutateFn := r.newFilesystemPVC(instance)
+
+	if result, err := controllerutil.CreateOrUpdate(ctx, r.Client, pvc, mutateFn); err != nil {
+		conditionsv1.SetStatusCondition(&instance.Status.Conditions, conditionsv1.Condition{
+			Type:    adiiov1alpha1.ConditionReconcileCompleted,
+			Status:  corev1.ConditionFalse,
+			Reason:  "StorageFailure",
+			Message: "Failed to ensure filesystem storage: " + err.Error(),
+		})
+		return err
+	} else if result != controllerutil.OperationResultNone {
+		r.Log.Info("Filesystem storage created")
+	}
+	return nil
+}
+
+func (r *AgentServiceConfigReconciler) ensureDatabaseStorage(ctx context.Context, instance *adiiov1alpha1.AgentServiceConfig) error {
+	pvc, mutateFn := r.newDatabasePVC(instance)
+
+	if result, err := controllerutil.CreateOrUpdate(ctx, r.Client, pvc, mutateFn); err != nil {
+		conditionsv1.SetStatusCondition(&instance.Status.Conditions, conditionsv1.Condition{
+			Type:    adiiov1alpha1.ConditionReconcileCompleted,
+			Status:  corev1.ConditionFalse,
+			Reason:  "StorageFailure",
+			Message: "Failed to ensure database storage: " + err.Error(),
+		})
+		return err
+	} else if result != controllerutil.OperationResultNone {
+		r.Log.Info("Database storage created")
+	}
+	return nil
+}
+
+func (r *AgentServiceConfigReconciler) ensureAgentService(ctx context.Context, instance *adiiov1alpha1.AgentServiceConfig) error {
+	svc, mutateFn := r.newAgentService(instance)
+
+	if result, err := controllerutil.CreateOrUpdate(ctx, r.Client, svc, mutateFn); err != nil {
+		conditionsv1.SetStatusCondition(&instance.Status.Conditions, conditionsv1.Condition{
+			Type:    adiiov1alpha1.ConditionReconcileCompleted,
+			Status:  corev1.ConditionFalse,
+			Reason:  "AgentServiceFailure",
```

Comment:
Probably a good idea to collect these reasons as constants with a doc string so there will be some reference for them.